src/server/game/Entities/AreaTrigger/AreaTrigger.cpp

```diff
@@ -100,14 +148,258 @@ void AreaTrigger::Update(uint32 p_time)
+    VisitNearbyObject(GetTemplate()->MaxSearchRadius, searcher);
+
+    HandleUnitEnterExit(targetList);
+}
+
+void AreaTrigger::SearchUnitInBox()
+{
+    float extentsX = GetTemplate()->BoxDatas.Extents[0];
+    float extentsY = GetTemplate()->BoxDatas.Extents[1];
+    float extentsZ = GetTemplate()->BoxDatas.Extents[2];
+
+    std::list<Unit*> targetList;
+
+    Trinity::AnyUnitInObjectRangeCheck check(this, GetTemplate()->MaxSearchRadius);
+    Trinity::UnitListSearcher<Trinity::AnyUnitInObjectRangeCheck> searcher(this, targetList, check);
+    VisitNearbyObject(GetTemplate()->MaxSearchRadius, searcher);
+
+    float halfExtentsX = extentsX / 2.0;
+    float halfExtentsY = extentsY / 2.0;
+    float halfExtentsZ = extentsZ / 2.0;
+
+    float minX = GetPositionX() - halfExtentsX;
+    float maxX = GetPositionX() + halfExtentsX;
+
+    float minY = GetPositionY() - halfExtentsY;
+    float maxY = GetPositionY() + halfExtentsY;
+
+    float minZ = GetPositionZ() - halfExtentsZ;
+    float maxZ = GetPositionZ() + halfExtentsZ;
+
+    G3D::AABox box({ minX, minY, minZ }, { maxX, maxY, maxZ });
+
+    targetList.remove_if([this, box](Unit* unit) -> bool
+    {
+        return !box.contains({ unit->GetPositionX(), unit->GetPositionY(), unit->GetPositionZ() });
+    });
+
+    HandleUnitEnterExit(targetList);
+}
+
+void AreaTrigger::SearchUnitInPolygon()
+{
+    std::list<Unit*> targetList;
+
+    Trinity::AnyUnitInObjectRangeCheck check(this, GetTemplate()->MaxSearchRadius);
+    Trinity::UnitListSearcher<Trinity::AnyUnitInObjectRangeCheck> searcher(this, targetList, check);
+    VisitNearbyObject(GetTemplate()->MaxSearchRadius, searcher);
+
+    float height = GetTemplate()->PolygonDatas.Height;
+
+    targetList.remove_if([this, height](Unit* unit) -> bool
+    {
+        float minZ = GetPositionZ() - height;
+        float maxZ = GetPositionZ() + height;
+
+        return  !CheckIsInPolygon2D(unit) ||
+                unit->GetPositionZ() < minZ ||
+                unit->GetPositionZ() > maxZ;
+    });
+
+    HandleUnitEnterExit(targetList);
+}
+
+void AreaTrigger::SearchUnitInCylinder()
+{
+    float radius = GetTemplate()->CylinderDatas.Radius;
+    float height = GetTemplate()->CylinderDatas.Height;
+
+    std::list<Unit*> targetList;
+
+    Trinity::AnyUnitInObjectRangeCheck check(this, GetTemplate()->MaxSearchRadius);
+    Trinity::UnitListSearcher<Trinity::AnyUnitInObjectRangeCheck> searcher(this, targetList, check);
+    VisitNearbyObject(GetTemplate()->MaxSearchRadius, searcher);
+
+    targetList.remove_if([this, radius, height](Unit* unit) -> bool
+    {
+        float minZ = GetPositionZ() - height;
+        float maxZ = GetPositionZ() + height;
+
+        return GetDistance(unit) > radius || unit->GetPositionZ() < minZ || unit->GetPositionZ() > maxZ;
+    });
+
+    HandleUnitEnterExit(targetList);
+}
+
+void AreaTrigger::HandleUnitEnterExit(std::list<Unit*> newTargetList)
+{
+    for (Unit* unit : newTargetList)
```

Comment:
I believe this would be a better way to handle this:
```
GuidUnorderedSet exitUnits = _insideUnits;
_insideUnits.clear();
for (Unit* unit : newTargetList)
{
    if (exitUnits.erase(unit) == 0) // erase(key_type) returns number of elements erased
    {
        // handle entering, chat handler and script hook
    }
    _insideUnits.insert(unit->GetGUID());
}

for (ObjectGuid const& exitUnitGuid : exitUnits)
{
    // handle exit, no conditions
}
```
And you don't need to check IsAlive() anywhere, AnyUnitInObjectRangeCheck does that so `newTargetList` will never contain dead units